runtime/gc_realtime/RealtimeAccessBarrier.hpp

```diff
@@ -97,18 +100,66 @@ class MM_RealtimeAccessBarrier : public MM_ObjectAccessBarrier
 	void validateWriteBarrier(J9VMThread *vmThread, J9Object *dstObject, fj9object_t *dstAddress, J9Object *srcObject);
-	virtual void rememberObjectImpl(MM_EnvironmentBase *env, J9Object *object) = 0;
-	
+	virtual void rememberObjectImpl(MM_EnvironmentBase *env, J9Object *object);
+
 	/* New methods */
 	void rememberObject(MM_EnvironmentBase *env, J9Object *object);
 	void rememberObjectIfBarrierEnabled(J9VMThread *vmThread, J9Object* object);
-	
+
+public:
+	static MM_RealtimeAccessBarrier *newInstance(MM_EnvironmentBase *env);
+	MMINLINE void setDoubleBarrierActive() { _doubleBarrierActive = true; }
+	MMINLINE void setDoubleBarrierInactive() { _doubleBarrierActive = false; }
+	MMINLINE bool isDoubleBarrierActive() { return _doubleBarrierActive; }
+	void setDoubleBarrierActiveOnThread(MM_EnvironmentBase* env);
+	void setDoubleBarrierInactiveOnThread(MM_EnvironmentBase* env);
+	virtual void initializeForNewThread(MM_EnvironmentBase* env);
+
+	virtual bool preObjectStore(J9VMThread *vmThread, J9Object *destObject, fj9object_t *destAddress, J9Object *value, bool isVolatile=false);
+	virtual bool preObjectStore(J9VMThread *vmThread, J9Object *destClass, J9Object **destAddress, J9Object *value, bool isVolatile=false);
+	virtual bool preObjectStore(J9VMThread *vmThread, J9Object **destAddress, J9Object *value, bool isVolatile=false);
+
+	virtual I_32 backwardReferenceArrayCopyIndex(J9VMThread *vmThread, J9IndexableObject *srcObject, J9IndexableObject *destObject, I_32 srcIndex, I_32 destIndex, I_32 lengthInSlots);
+	virtual I_32 forwardReferenceArrayCopyIndex(J9VMThread *vmThread, J9IndexableObject *srcObject, J9IndexableObject *destObject, I_32 srcIndex, I_32 destIndex, I_32 lengthInSlots);
+
+	bool preObjectStoreInternal(J9VMThread *vmThread, J9Object *destClass, J9Object **destAddress, J9Object *value, bool isVolatile);
+	bool preObjectStoreInternal(J9VMThread *vmThread, J9Object *destObject, fj9object_t *destAddress, J9Object *value, bool isVolatile);
+	bool preObjectStoreInternal(J9VMThread *vmThread, J9Object **destAddress, J9Object *value, bool isVolatile);
+
+	MMINLINE bool isBarrierActive(MM_EnvironmentBase* env)
+	{
+		MM_GCExtensions* extensions = MM_GCExtensions::getExtensions(env);
+		return !extensions->sATBBarrierRememberedSet->isGlobalFragmentIndexPreserved(env);
+	}
+
+	MMINLINE bool isDoubleBarrierActiveOnThread(J9VMThread *vmThread)
+	{
+		/* The double barrier is enabled in staccato by setting the threads remembered set fragment index
+		 * to the special value, this ensures the JIT will go out-of line. We can determine if the double
+		 * barrier is active simply by checking if the fragment index corresponds to the special value.
+		 */
+		return (J9GC_REMEMBERED_SET_RESERVED_INDEX == vmThread->sATBBarrierRememberedSetFragment.localFragmentIndex);
+	}
+
+	bool markAndScanContiguousArray(MM_EnvironmentRealtime *env, J9IndexableObject *objectPtr);
+	void scanContiguousArray(MM_EnvironmentRealtime *env, J9IndexableObject *objectPtr);
```

Comment:
These two methods use to be protected. Looks like it might be private now. At the moment they are declared public.
I would merge it "as is" but we need clean up it later